Error with genrics and lambda

I have the following code

```cs
        [Zomp.SyncMethodGenerator.CreateSyncVersion(OmitNullableDirective = true)]
        public async Task<bool> FooAsync()
        {
            return await Bar(() => Task.FromResult(true));
        }

        public T Bar<T>(Func<T> innerLogic) => innerLogic();
```

which ends up producing 

```cs
        public bool Foo()
        {
        }
```

which gives the following error 

```
FooAsync.g.cs(6,21,6,24): error CS0161: 'Foo()': not all code paths return a value
```

I expected it to produce

```cs
        public bool Foo()
        {
             return Bar(() => true);
        }
```

--- a/src/Zomp.SyncMethodGenerator/AsyncToSyncRewriter.cs
+++ b/src/Zomp.SyncMethodGenerator/AsyncToSyncRewriter.cs
@@ -863,10 +863,14 @@ internal sealed class AsyncToSyncRewriter(SemanticModel semanticModel, bool disa
         var isFunc = node.Parent is { } parent
             && semanticModel.GetTypeInfo(parent).Type is INamedTypeSymbol { IsGenericType: true, IsSystemFunc: true } n;
 
+        // Generic method type arguments like Bar<Task<T>> are unwrapped to Bar<T>, not removed
+        var isMethodTypeArgumentList = node.Parent is GenericNameSyntax gn && GetSymbol(gn) is IMethodSymbol;
+
         // Do not remove Task<T>, but remove Task inside a Func<>
         bool RemoveType(TypeSyntax z, int index) =>
             !(isFunc && index == node.Arguments.Count - 1 && z is GenericNameSyntax)
             && semanticModel.GetTypeInfo(z).Type is { } type
+            && !(isMethodTypeArgumentList && type is INamedTypeSymbol { IsTaskOrValueTask: true, IsGenericType: true })
             && ShouldRemoveType(type);
 
         var indicesToRemove = node.Arguments.GetIndices(RemoveType);
@@ -1917,7 +1921,8 @@ internal sealed class AsyncToSyncRewriter(SemanticModel semanticModel, bool disa
         IPropertySymbol { IsCancellationRequested: true } => !isNegated,
         IMethodSymbol ms =>
             IsSpecialMethod(ms) is SpecialMethod.None or SpecialMethod.Drop
-                && ((ShouldRemoveType(ms.ReturnType, isArgument: true) && ms.MethodKind != MethodKind.LocalFunction)
+                && ((ShouldRemoveType(ms.ReturnType, isArgument: true) && ms.MethodKind != MethodKind.LocalFunction
+                        && ms.OriginalDefinition.ReturnType is not ITypeParameterSymbol { TypeParameterKind: TypeParameterKind.Method })
                     || (ms.ReceiverType is { } receiver && ShouldRemoveType(receiver, isArgument: true)))
                 && !HasSyncMethod(ms),
         _ => ShouldRemoveType(GetReturnType(symbol), isArgument: true),

--- a/tests/Generator.Tests/Snapshots/UnitTests.GenericMethodInferringTaskTypeArgument#FooAsync.g.verified.cs
+++ b/tests/Generator.Tests/Snapshots/UnitTests.GenericMethodInferringTaskTypeArgument#FooAsync.g.verified.cs
@@ -1,0 +1,5 @@
+﻿//HintName: Test.Class.FooAsync.g.cs
+public bool Foo()
+{
+    return Bar(() => true);
+}

--- a/tests/Generator.Tests/Snapshots/UnitTests.GenericMethodWithExplicitTaskTypeArgument#FooAsync.g.verified.cs
+++ b/tests/Generator.Tests/Snapshots/UnitTests.GenericMethodWithExplicitTaskTypeArgument#FooAsync.g.verified.cs
@@ -1,0 +1,5 @@
+﻿//HintName: Test.Class.FooAsync.g.cs
+public bool Foo()
+{
+    return Bar<bool>(() => true);
+}

--- a/tests/Generator.Tests/UnitTests.cs
+++ b/tests/Generator.Tests/UnitTests.cs
@@ -53,6 +53,28 @@ namespace M
 """.Verify(sourceType: SourceType.Full);
 
     [Fact]
+    public Task GenericMethodInferringTaskTypeArgument() => """
+[CreateSyncVersion]
+public async Task<bool> FooAsync()
+{
+    return await Bar(() => Task.FromResult(true));
+}
+
+public T Bar<T>(Func<T> innerLogic) => innerLogic();
+""".Verify();
+
+    [Fact]
+    public Task GenericMethodWithExplicitTaskTypeArgument() => """
+[CreateSyncVersion]
+public async Task<bool> FooAsync()
+{
+    return await Bar<Task<bool>>(() => Task.FromResult(true));
+}
+
+public T Bar<T>(Func<T> innerLogic) => innerLogic();
+""".Verify();
+
+    [Fact]
     public Task MultipleNamespaces() => """
 namespace NsOne
 {

